File: examples/wasm-scanner/src/scanner-shell.mjs
import { GrainSnapshotCoordinator } from "../../../sdk/wasm/src/browser-storage.mjs";

export const SCANNER_ACCEPT_REQUIRES_VERIFIED_PREVIEW_DIAG =
  "SDK_ERR_EXAMPLE_ACCEPT_REQUIRES_VERIFIED_PREVIEW";
export const SCANNER_SNAPSHOT_PERSISTENCE_DIAG = "SDK_ERR_EXAMPLE_SNAPSHOT_PERSISTENCE";

export function createScannerShell(client, { trustProvider, snapshotPersistence = null } = {}) {
  requireClient(client);
  requireTrustProvider(trustProvider);
  const snapshotCoordinator = snapshotPersistence === null ? null : new GrainSnapshotCoordinator(snapshotPersistence);

  const state = {
    qrString: "",
    trustAnchorId: "",
    previewStatus: null,
    acceptStatus: null,
    diagnostics: [],
    canAccept: false,
    acceptedCount: 0,
    acceptedScanId: null,
    lifecycleStatus: null,
    deviceCount: 0,
    lifecycleEventCount: 0,
    snapshotStatus: null,
    exportStatus: null,
    exportAcceptedCount: 0,
    exportDeviceCount: 0,
    exportLifecycleEventCount: 0,
  };

  return {
    get state() {
      return structuredClone(state);
    },

    setQrString(value) {
      state.qrString = String(value);
      resetDecisionState(state);
    },

    setTrustAnchorId(value) {
      state.trustAnchorId = String(value);
      resetDecisionState(state);
    },

    receiveCameraPayload(payload) {
      state.qrString = requireCameraPayload(payload).qrString;
      resetDecisionState(state);
    },

    async prepareLocalIdentity({ rootLabel = "phone", deviceLabel = "scanner" } = {}) {
      const lifecycle = client.clientLifecycle();
      if (lifecycle.status === "Ready") {
        state.diagnostics = lifecycle.diag;
        applyLifecycle(state, lifecycle);
        return this.state;
      }
      if (lifecycle.status === "Uninitialized") {
        const root = client.createRootIdentity({ label: rootLabel });
        if (root.diag.length > 0) {
          state.diagnostics = root.diag;
          refreshLifecycle(state, client);
          return this.state;
        }
      }

      const device = client.addDeviceKey({ label: deviceLabel });
      state.diagnostics = device.diag;
      refreshLifecycle(state, client);
      await persistSnapshot(state, client, snapshotCoordinator);
      return this.state;
    },

    preview() {
      const preview = client.scanPreviewWithTrustProvider({
        qrString: state.qrString,
        trustAnchorId: normalizedTrustAnchorId(state),
        trustProvider,
      });

      state.previewStatus = preview.status;
      state.acceptStatus = null;
      state.diagnostics = preview.diag;
      state.canAccept = preview.status === "Verified";
      state.acceptedCount = client.listAcceptedScans().length;
      state.acceptedScanId = null;
      return this.state;
    },

    async accept() {
      if (state.previewStatus !== "Verified" || !state.canAccept) {
        const diagnostics = state.previewStatus === "Rejected" && state.diagnostics.length > 0
          ? state.diagnostics
          : [SCANNER_ACCEPT_REQUIRES_VERIFIED_PREVIEW_DIAG];
        state.acceptStatus = null;
        state.diagnostics = diagnostics;
        state.canAccept = false;
        state.acceptedScanId = null;
        return this.state;
      }

      const accepted = client.scanAcceptWithTrustProvider({
        qrString: state.qrString,
        trustAnchorId: normalizedTrustAnchorId(state),
        trustProvider,
      });

      state.acceptStatus = accepted.status;
      state.acceptedScanId = accepted.scanId;
      state.diagnostics = accepted.diag;
      state.acceptedCount = client.listAcceptedScans().length;
      state.canAccept = state.previewStatus === "Verified";
      if (accepted.status === "Accepted" || accepted.status === "AlreadyAccepted") {
        await persistSnapshot(state, client, snapshotCoordinator);
      }
      return this.state;
    },

    async restorePersistedSnapshot() {
      if (snapshotCoordinator === null) {
        return this.state;
      }
      try {
        const restored = await snapshotCoordinator.restore(client);
        if (restored === null) {
          state.snapshotStatus = "Empty";
          refreshLifecycle(state, client);
          state.acceptedCount = client.listAcceptedScans().length;
          return this.state;
        }
        state.snapshotStatus = restored.status;
        state.diagnostics = restored.diag;
        state.acceptedCount = restored.acceptedRecordCount;
        refreshLifecycle(state, client);
      } catch (_) {
        state.snapshotStatus = "PersistenceError";
        state.diagnostics = [SCANNER_SNAPSHOT_PERSISTENCE_DIAG];
      }
      return this.state;
    },

    exportSyncBundleForShare() {
      const exported = client.exportSyncBundle();
      state.exportStatus = exported.status;
      state.exportAcceptedCount = exported.acceptedRecordCount;
      state.exportDeviceCount = exported.deviceCount;
      state.exportLifecycleEventCount = exported.lifecycleEventCount;
      state.diagnostics = exported.diag;
      return exported;
    },
  };
}

export function mountScannerShell(root, client, {
  cameraAdapter = null,
  trustProvider,
  snapshotPersistence = null,
} = {}) {
  const shell = createScannerShell(client, { trustProvider, snapshotPersistence });
  root.replaceChildren();
  root.classList.add("grain-scanner");

  const scanInput = document.createElement("textarea");
  scanInput.placeholder = "GR1 string";
  scanInput.rows = 6;

  const trustInput = document.createElement("input");
  trustInput.placeholder = "Trust anchor ID";
  trustInput.autocapitalize = "off";
  trustInput.autocomplete = "off";
  trustInput.spellcheck = false;

  const previewButton = document.createElement("button");
  previewButton.type = "button";
  previewButton.textContent = "Preview";

  const prepareButton = document.createElement("button");
  prepareButton.type = "button";
  prepareButton.textContent = "Prepare device";

  const acceptButton = document.createElement("button");
  acceptButton.type = "button";
  acceptButton.textContent = "Accept";
  acceptButton.disabled = true;

  const cameraButton = document.createElement("button");
  cameraButton.type = "button";
  cameraButton.textContent = "Camera";
  cameraButton.hidden = cameraAdapter === null;

  const restoreButton = document.createElement("button");
  restoreButton.type = "button";
  restoreButton.textContent = "Restore";
  restoreButton.hidden = snapshotPersistence === null;

  const video = document.createElement("video");
  video.playsInline = true;
  video.muted = true;
  video.hidden = cameraAdapter === null;

  const status = document.createElement("output");
  const diagnostics = document.createElement("ul");

  function render() {
    const current = shell.state;
    acceptButton.disabled = !current.canAccept;
    status.value = [
      current.previewStatus ? `Preview: ${current.previewStatus}` : null,
      current.acceptStatus ? `Accept: ${current.acceptStatus}` : null,
      current.lifecycleStatus ? `Lifecycle: ${current.lifecycleStatus}` : null,
      current.lifecycleStatus ? `Devices: ${current.deviceCount}` : null,
      current.snapshotStatus ? `Snapshot: ${current.snapshotStatus}` : null,
      current.exportStatus ? `Export: ${current.exportStatus}` : null,
      `Saved: ${current.acceptedCount}`,
    ].filter(Boolean).join(" | ");
    diagnostics.replaceChildren(...current.diagnostics.map((diagnostic) => {
      const item = document.createElement("li");
      item.textContent = diagnostic;
      return item;
    }));
  }

  scanInput.addEventListener("input", () => {
    shell.setQrString(scanInput.value);
    render();
  });
  trustInput.addEventListener("input", () => {
    shell.setTrustAnchorId(trustInput.value);
    render();
  });
  previewButton.addEventListener("click", () => {
    shell.preview();
    render();
  });
  prepareButton.addEventListener("click", async () => {
    await shell.prepareLocalIdentity();
    render();
  });
  acceptButton.addEventListener("click", async () => {
    await shell.accept();
    render();
  });
  restoreButton.addEventListener("click", async () => {
    await shell.restorePersistedSnapshot();
    render();
  });
  cameraButton.addEventListener("click", async () => {
    if (cameraAdapter.start) {
      await cameraAdapter.start(video);
    }
    const payload = await cameraAdapter.scanOnce(video);
    shell.receiveCameraPayload(payload);
    scanInput.value = payload.qrString;
    render();
  });

  root.append(
    scanInput,
    trustInput,
    prepareButton,
    restoreButton,
    previewButton,
    acceptButton,
    cameraButton,
    video,
    status,
    diagnostics,
  );
  render();
  return shell;
}

function requireClient(client) {
  for (const method of [
    "scanPreviewWithTrustProvider",
    "scanAcceptWithTrustProvider",
    "listAcceptedScans",
    "createRootIdentity",
    "addDeviceKey",
    "clientLifecycle",
    "exportSyncBundle",
  ]) {
    if (!client || typeof client[method] !== "function") {
      throw new TypeError(`SDK_ERR_EXAMPLE_CLIENT_METHOD_MISSING:${method}`);
    }
  }
}

function requireTrustProvider(trustProvider) {
  if (!trustProvider || typeof trustProvider.trustPubB64 !== "function") {
    throw new TypeError("SDK_ERR_EXAMPLE_TRUST_PROVIDER_MISSING");
  }
}

function refreshLifecycle(state, client) {
  const lifecycle = client.clientLifecycle();
  applyLifecycle(state, lifecycle);
}

function applyLifecycle(state, lifecycle) {
  state.lifecycleStatus = lifecycle.status;
  state.deviceCount = lifecycle.deviceCount;
  state.lifecycleEventCount = lifecycle.lifecycleEventCount;
}

function normalizedTrustAnchorId(state) {
  return state.trustAnchorId.trim();
}

async function persistSnapshot(state, client, snapshotCoordinator) {
  if (snapshotCoordinator === null) {
    return;
  }
  try {
    const exported = await snapshotCoordinator.persist(client);
    state.snapshotStatus = exported.status;
    if (exported.diag.length > 0) {
      state.diagnostics = exported.diag;
    }
  } catch (_) {
    state.snapshotStatus = "PersistenceError";
    state.diagnostics = [SCANNER_SNAPSHOT_PERSISTENCE_DIAG];
  }
}

function requireCameraPayload(payload) {
  if (!payload || typeof payload.qrString !== "string" || payload.qrString.length === 0) {
    throw new TypeError("SDK_ERR_EXAMPLE_CAMERA_PAYLOAD_INVALID");
  }
  return payload;
}

function resetDecisionState(state) {
  state.previewStatus = null;
  state.acceptStatus = null;
  state.diagnostics = [];
  state.canAccept = false;
  state.acceptedScanId = null;
  state.exportStatus = null;
  state.exportAcceptedCount = 0;
  state.exportDeviceCount = 0;
  state.exportLifecycleEventCount = 0;
}

File: examples/android-scanner/src/main/kotlin/dev/grain/examples/androidscanner/ScannerShell.kt
package dev.grain.examples.androidscanner

import dev.grain.GrainAcceptedScan
import dev.grain.GrainClient
import dev.grain.GrainClientLifecycle
import dev.grain.GrainDeviceResult
import dev.grain.GrainIdentityResult
import dev.grain.GrainScanAccept
import dev.grain.GrainScanAcceptStatus
import dev.grain.GrainScanPreview
import dev.grain.GrainScanPreviewStatus
import dev.grain.GrainStoreSnapshotResult
import dev.grain.GrainSyncResult
import dev.grain.GrainTrustProvider
import dev.grain.android.GrainSnapshotClient
import dev.grain.android.GrainSnapshotCoordinator
import dev.grain.android.GrainSnapshotPersistence

const val SCANNER_ACCEPT_REQUIRES_VERIFIED_PREVIEW_DIAG =
    "SDK_ERR_EXAMPLE_ACCEPT_REQUIRES_VERIFIED_PREVIEW"
const val SCANNER_SNAPSHOT_PERSISTENCE_DIAG = "SDK_ERR_EXAMPLE_SNAPSHOT_PERSISTENCE"

interface ScannerWorkflowClient : GrainSnapshotClient {
    fun scanPreview(
        qrString: String,
        trustAnchorId: String,
        trustProvider: GrainTrustProvider,
    ): GrainScanPreview
    fun scanAccept(
        qrString: String,
        trustAnchorId: String,
        trustProvider: GrainTrustProvider,
    ): GrainScanAccept
    fun listAcceptedScans(): List<GrainAcceptedScan>
    fun exportSyncBundle(): GrainSyncResult
    fun createRootIdentity(label: String = "root"): GrainIdentityResult
    fun addDeviceKey(label: String = "device"): GrainDeviceResult
    fun clientLifecycle(): GrainClientLifecycle
}

class GrainScannerWorkflowClient(
    private val client: GrainClient = GrainClient(),
) : ScannerWorkflowClient, AutoCloseable {
    override fun scanPreview(
        qrString: String,
        trustAnchorId: String,
        trustProvider: GrainTrustProvider,
    ): GrainScanPreview =
        client.scanPreview(
            qrString = qrString,
            trustAnchorId = trustAnchorId,
            trustProvider = trustProvider,
        )

    override fun scanAccept(
        qrString: String,
        trustAnchorId: String,
        trustProvider: GrainTrustProvider,
    ): GrainScanAccept =
        client.scanAccept(
            qrString = qrString,
            trustAnchorId = trustAnchorId,
            trustProvider = trustProvider,
        )

    override fun listAcceptedScans(): List<GrainAcceptedScan> =
        client.listAcceptedScans()

    override fun exportSyncBundle(): GrainSyncResult =
        client.exportSyncBundle()

    override fun createRootIdentity(label: String): GrainIdentityResult =
        client.createRootIdentity(label = label)

    override fun addDeviceKey(label: String): GrainDeviceResult =
        client.addDeviceKey(label = label)

    override fun clientLifecycle(): GrainClientLifecycle =
        client.clientLifecycle()

    override fun exportStoreSnapshot(): GrainStoreSnapshotResult =
        client.exportStoreSnapshot()

    override fun restoreStoreSnapshot(snapshotB64: String): GrainStoreSnapshotResult =
        client.restoreStoreSnapshot(snapshotB64 = snapshotB64)

    override fun close() {
        client.close()
    }
}

data class ScannerAcceptedScanSummary(
    val scanId: String,
)

data class ScannerUiState(
    val qrString: String = "",
    val trustAnchorId: String = "",
    val previewStatus: GrainScanPreviewStatus? = null,
    val acceptStatus: GrainScanAcceptStatus? = null,
    val scanSource: CameraScanSource? = null,
    val diagnostics: List<String> = emptyList(),
    val canAccept: Boolean = false,
    val acceptedCount: Int = 0,
    val acceptedScans: List<ScannerAcceptedScanSummary> = emptyList(),
    val acceptedScanId: String? = null,
    val lifecycleStatus: String? = null,
    val deviceCount: ULong = 0UL,
    val lifecycleEventCount: ULong = 0UL,
    val snapshotStatus: String? = null,
    val exportStatus: String? = null,
    val exportAcceptedCount: ULong = 0UL,
)

class ScannerController(
    private val client: ScannerWorkflowClient,
    private val trustProvider: GrainTrustProvider,
    snapshotPersistence: GrainSnapshotPersistence? = null,
) {
    private val snapshotCoordinator = snapshotPersistence?.let(::GrainSnapshotCoordinator)
    var state: ScannerUiState = ScannerUiState()
        private set

    fun updateQrString(value: String) {
        state = state.copy(qrString = value, scanSource = null).withoutDecision()
    }

    fun updateTrustAnchorId(value: String) {
        state = state.copy(trustAnchorId = value).withoutDecision()
    }

    fun receiveCameraPayload(payload: CameraScanPayload) {
        state = state.copy(qrString = payload.qrString, scanSource = payload.source).withoutDecision()
    }

    fun <FrameT> scanCameraFrame(
        frame: FrameT,
        cameraScanAdapter: CameraScanAdapter<FrameT>,
    ): Boolean {
        val payload = cameraScanAdapter.decode(frame) ?: return false
        receiveCameraPayload(payload)
        return true
    }

    fun prepareLocalIdentity(rootLabel: String = "phone", deviceLabel: String = "scanner") {
        val lifecycle = client.clientLifecycle()
        if (lifecycle.status == "Ready") {
            state = state.copy(diagnostics = lifecycle.diag).withLifecycle(lifecycle)
            return
        }
        if (lifecycle.status == "Uninitialized") {
            val root = client.createRootIdentity(label = rootLabel)
            if (root.diag.isNotEmpty()) {
                state = state.copy(diagnostics = root.diag).withLifecycle(client.clientLifecycle())
                return
            }
        }

        val device = client.addDeviceKey(label = deviceLabel)
        state = state
            .copy(diagnostics = device.diag)
            .withLifecycle(client.clientLifecycle())
        persistSnapshot()
    }

    fun preview() {
        val preview = client.scanPreview(
            qrString = state.qrString,
            trustAnchorId = normalizedTrustAnchorId(),
            trustProvider = trustProvider,
        )

        state = state.copy(
            previewStatus = preview.status,
            acceptStatus = null,
            diagnostics = preview.diag,
            canAccept = preview.status == GrainScanPreviewStatus.Verified,
            acceptedScanId = null,
        ).withAcceptedScans(client.listAcceptedScans())
    }

    fun accept() {
        if (state.previewStatus != GrainScanPreviewStatus.Verified || !state.canAccept) {
            val diagnostics = if (
                state.previewStatus == GrainScanPreviewStatus.Rejected &&
                state.diagnostics.isNotEmpty()
            ) {
                state.diagnostics
            } else {
                listOf(SCANNER_ACCEPT_REQUIRES_VERIFIED_PREVIEW_DIAG)
            }
            state = state.copy(
                acceptStatus = null,
                diagnostics = diagnostics,
                canAccept = false,
                acceptedScanId = null,
            )
            return
        }

        val accepted = client.scanAccept(
            qrString = state.qrString,
            trustAnchorId = normalizedTrustAnchorId(),
            trustProvider = trustProvider,
        )
        val acceptedScans = client.listAcceptedScans()
        state = state.copy(
            acceptStatus = accepted.status,
            diagnostics = accepted.diag,
            acceptedScanId = accepted.scanId,
            canAccept = state.previewStatus == GrainScanPreviewStatus.Verified,
        ).withAcceptedScans(acceptedScans)
        if (
            accepted.status == GrainScanAcceptStatus.Accepted ||
            accepted.status == GrainScanAcceptStatus.AlreadyAccepted
        ) {
            persistSnapshot()
        }
    }

    fun refreshAcceptedScans() {
        state = state.withAcceptedScans(client.listAcceptedScans())
    }

    fun exportSyncBundleForShare(): GrainSyncResult {
        val exported = client.exportSyncBundle()
        state = state.copy(
            exportStatus = exported.status,
            exportAcceptedCount = exported.acceptedRecordCount,
            acceptedCount = exported.acceptedRecordCount.toInt(),
            diagnostics = if (exported.diag.isEmpty()) state.diagnostics else exported.diag,
        )
        return exported
    }

    fun restorePersistedSnapshot() {
        val coordinator = snapshotCoordinator ?: return
        try {
            val restored = coordinator.restore(client = client)
            if (restored == null) {
                state = state.copy(snapshotStatus = "Empty")
                refreshLifecycle()
                refreshAcceptedScans()
                return
            }
            state = state.copy(
                snapshotStatus = restored.status,
                diagnostics = restored.diag,
            )
            refreshLifecycle()
            refreshAcceptedScans()
        } catch (_: RuntimeException) {
            state = state.copy(
                snapshotStatus = "PersistenceError",
                diagnostics = listOf(SCANNER_SNAPSHOT_PERSISTENCE_DIAG),
            )
        }
    }

    private fun normalizedTrustAnchorId(): String =
        state.trustAnchorId.trim()

    private fun persistSnapshot() {
        val coordinator = snapshotCoordinator ?: return
        try {
            val exported = coordinator.persist(client = client)
            state = state.copy(
                snapshotStatus = exported.status,
                diagnostics = if (exported.diag.isEmpty()) state.diagnostics else exported.diag,
            )
        } catch (_: RuntimeException) {
            state = state.copy(
                snapshotStatus = "PersistenceError",
                diagnostics = listOf(SCANNER_SNAPSHOT_PERSISTENCE_DIAG),
            )
        }
    }

    private fun refreshLifecycle() {
        state = state.withLifecycle(client.clientLifecycle())
    }

    private fun ScannerUiState.withAcceptedScans(
        acceptedScans: List<GrainAcceptedScan>,
    ): ScannerUiState =
        copy(
            acceptedCount = acceptedScans.size,
            acceptedScans = acceptedScans.map { ScannerAcceptedScanSummary(scanId = it.scanId) },
        )

    private fun ScannerUiState.withoutDecision(): ScannerUiState =
        copy(
            previewStatus = null,
            acceptStatus = null,
            diagnostics = emptyList(),
            canAccept = false,
            acceptedScanId = null,
            exportStatus = null,
            exportAcceptedCount = 0UL,
        )

    private fun ScannerUiState.withLifecycle(lifecycle: GrainClientLifecycle): ScannerUiState =
        copy(
            lifecycleStatus = lifecycle.status,
            deviceCount = lifecycle.deviceCount,
            lifecycleEventCount = lifecycle.lifecycleEventCount,
        )
}
